bool[] contacts
duration time_from_start